hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlClientTest.java

```diff
@@ -108,4 +117,42 @@ public void when_resultClosed_then_jobCancelled_withNoResults() {
         logger.info("after res.close() returned");
         assertJobStatusEventually(job, FAILED);
     }
+
+    // test for https://github.com/hazelcast/hazelcast/issues/19897
+    @Test
+    public void when_resultClosed_then_executionContextCleanedUp() {
+        HazelcastInstance client = factory().newHazelcastClient();
+        SqlService sql = client.getSql();
+
+        IMap<Integer, Integer> map = instance().getMap("map");
+        Map<Integer, Integer> tmpMap = new HashMap<>();
+        for (int i = 0; i < 100_000; i++) {
+            tmpMap.put(i, i);
+            if (i % 10_000 == 0) {
+                map.putAll(tmpMap);
+                tmpMap.clear();
+            }
+        }
+
+        createMapping("map", Integer.class, Integer.class);
+
+        for (int i = 0; i < 100; i++) {
+            SqlResult result = sql.execute("SELECT * FROM map");
+            result.close();
+        }
+
+        JetServiceBackend jetService = getJetServiceBackend(instance());
+        Collection<ExecutionContext> contexts = jetService.getJobExecutionService().getExecutionContexts();
+        // assert that all ExecutionContexts are eventually cleaned up
+        assertTrueEventually(() -> {
+            String remainingContexts = contexts.stream()
+                    .map(c -> idToString(c.executionId()))
+                    .collect(Collectors.joining(", "));
+            assertEquals("remaining execIds: " + remainingContexts, 0, contexts.size());
+        }, 5);
+
+        // assert that failedJobs is also cleaned up
+        ConcurrentMap<Long, Long> failedJobs = jetService.getJobExecutionService().getFailedJobs();
+        assertTrueEventually(() -> assertEquals(0, failedJobs.size()));
```

Comment:
If you are worried about the test duration you could move this assertion to the previous `assertTrueEventually` block and increase the timeout there from 5 to some higher number (not sure what the exact number would be, the cleanup job runs every 1 second, the failedJobs has TTL 5 seconds, maybe there is something else that could cause a delay).

But because this test is also `ParallelJVMTest` I think it's ok as is.

Reply:
That will not make the test faster, the time is measured since the last job fails. It's not triggered by the test.